code/controllers/subsystem/materials.dm

```diff
@@ -31,40 +35,130 @@ SUBSYSTEM_DEF(materials)
+ *       - If the material type is not bespoke the type is assumed to be the id for a material and the corresponding material is loaded from the cache.
+ *       - If the material type is bespoke a text ID is generated from the arguments list and used to load a material datum from the cache.
+ *   - The following elements are used to generate bespoke IDs
+ */
+/datum/controller/subsystem/materials/proc/_GetMaterialRef(list/arguments)
 	if(!materials)
 		InitializeMaterials()
-	return materials[fakemat] || fakemat
 
-///Returns a list to be used as an object's custom_materials. Lists will be cached and re-used based on the parameters.
+	var/datum/material/key = arguments[1]
+	if(istype(key))
+		return key // We are assuming here that the only thing allowed to create material datums is [/datum/controller/subsystem/materials/proc/InitializeMaterial]
+
+	if(istext(key)) // Handle text id
+		. = materials[key]
+		if(!.)
+			WARNING("Attempted to fetch material ref with invalid text id '[key]'")
+		return
+
+	if(!ispath(key, /datum/material))
+		CRASH("Attempted to fetch material ref with invalid key [key]")
+
+	if(!(initial(key.init_flags) & MATERIAL_INIT_BESPOKE))
+		. = materials[key]
+		if(!.)
+			WARNING("Attempted to fetch reference to an abstract material with key [key]")
+		return
+
+	key = GetIdFromArguments(arguments)
+	return materials[key] || InitializeMaterial(arguments)
+
+/** I'm not going to lie, this was swiped from [SSdcs][/datum/controller/subsystem/processing/dcs].
+ * Credit does to ninjanomnom
+ *
+ * Generates an id for bespoke ~~elements~~ materials when given the argument list
+ * Generating the id here is a bit complex because we need to support named arguments
+ * Named arguments can appear in any order and we need them to appear after ordered arguments
+ * We assume that no one will pass in a named argument with a value of null
+ **/
+/datum/controller/subsystem/materials/proc/GetIdFromArguments(list/arguments)
+	var/datum/material/mattype = arguments[1]
+	var/list/fullid = list("[initial(mattype.id) || mattype]")
+	var/list/named_arguments = list()
+	for(var/i in 2 to length(arguments))
+		var/key = arguments[i]
+		var/value
+		if(istext(key))
+			value = arguments[key]
+		if(!(istext(key) || isnum(key)))
+			key = REF(key)
+		key = "[key]" // Key is stringified so numbers dont break things
+		if(!isnull(value))
+			if(!(istext(value) || isnum(value)))
+				value = REF(value)
+			named_arguments["[key]"] = value
+		else
+			fullid += "[key]"
+
+	if(length(named_arguments))
+		named_arguments = sortList(named_arguments)
+		fullid += named_arguments
+	return list2params(fullid)
+
+
+/// Returns a list to be used as an object's custom_materials. Lists will be cached and re-used based on the parameters.
 /datum/controller/subsystem/materials/proc/FindOrCreateMaterialCombo(list/materials_declaration, multiplier)
+	if(isnull(materials_declaration))
+		return // Shouldn't happen. Will happen.
```

Comment:
throw an error then instead

Reply:
Shouldn't this just pass back a null?

Reply:
if this happens it means theres something going so doing CRASH() will make sure we can figure out why rather than having silent errors.

Reply:
The problem is that there are legitimate reasons to pass this a null. Mostly clearing out somethings custom materials.

Reply:
then the comment is wrong. it literaly says it shouldnt happen.

Reply:
Fixed the comment